storage/redis/peer_store.go

```diff
@@ -0,0 +1,803 @@
+		})
+		return
+	}
+	sLen := seedersLen.(int64)
+
+	if lLen == 0 && sLen == 0 {
+		return
+	}
+
+	resp.Incomplete = uint32(lLen)
+	resp.Complete = uint32(sLen)
+
+	return
+}
+
+// collectGarbage deletes all Peers from the PeerStore which are older than the
+// cutoff time.
+//
+// This function must be able to execute while other methods on this interface
+// are being executed in parallel.
+func (ps *peerStore) collectGarbage(cutoff time.Time) error {
+	select {
+	case <-ps.closed:
+		return nil
+	default:
+	}
+
+	shards := [2]string{bittorrent.IPv4.String(), bittorrent.IPv6.String()}
+
+	conn := ps.rb.open()
+	defer conn.Close()
+
+	cutoffUnix := cutoff.UnixNano()
+	start := time.Now()
+
+	for _, shard := range shards {
+		infohashesList, err := conn.Do("HKEYS", shard) // key
+		if err != nil {
+			return err
+		}
+		infohashes := infohashesList.([]interface{})
+
+		for _, ih := range infohashes {
+			ihStr := string(ih.([]byte))
+
+			ihList, err := conn.Do("HGETALL", ihStr) // field
+			if err != nil {
+				return err
+			}
+			conIhList := ihList.([]interface{})
+
+			if len(conIhList) == 0 {
+				_, err := conn.Do("DEL", ihStr)
+				if err != nil {
+					return err
+				}
+				log.Debug("storage: Deleting Redis", log.Fields{"Hkey": ihStr})
+				_, err = conn.Do("HDEL", shard, ihStr)
+				if err != nil {
+					return err
+				}
+				log.Debug("storage: Deleting Redis", log.Fields{
+					"Hkey":   shard,
+					"Hfield": ihStr,
+				})
+				continue
+			}
+
+			var pk serializedPeer
+			for index, ihField := range conIhList {
+				if index%2 != 0 { // value
+					mtime, err := strconv.ParseInt(string(ihField.([]byte)), 10, 64)
+					if err != nil {
+						return err
+					}
+					if mtime <= cutoffUnix {
+						_, err := conn.Do("HDEL", ihStr, pk)
+						if err != nil {
+							return err
+						}
+						p := decodePeerKey(pk)
+						log.Debug("storage: Deleting peer", log.Fields{
+							"Peer": fmt.Sprintf("[ID: %s, IP: %s(AddressFamily: %s), Port %d]", p.ID.String(), p.IP.String(), p.IP.AddressFamily.String(), p.Port),
+						})
+					}
+				} else { // key
+					pk = serializedPeer(ihField.([]byte))
+				}
+			}
+
+			ihLen, err := conn.Do("HLEN", ihStr)
+			if err != nil {
+				return err
+			}
+			if ihLen.(int64) == 0 {
+				_, err := conn.Do("DEL", ihStr)
+				if err != nil {
+					return err
+				}
+				log.Debug("storage: Deleting Redis", log.Fields{"Hkey": ihStr})
+				_, err = conn.Do("HDEL", shard, ihStr)
+				if err != nil {
+					return err
+				}
+				log.Debug("storage: Deleting Redis", log.Fields{
+					"Hkey":   shard,
+					"Hfield": ihStr,
+				})
+			}
+
+		}
+
+	}
+
+	recordGCDuration(time.Since(start))
+
+	return nil
+}
+
+func (ps *peerStore) Stop() stop.Result {
+	c := make(stop.Channel)
+	go func() {
+		close(ps.closed)
+		ps.wg.Wait()
+		ps.clean()
+		close(c)
+	}()
+
+	return c.Result()
+}
+
+func (ps *peerStore) LogFields() log.Fields {
+	return ps.cfg.LogFields()
+}
+
+func (ps *peerStore) clean() {
```

Comment:
Oh, this is an interesting question: Do we really want to clear redis when we shut down? Imagine you have two instances of chihaya running on two machines, now for some reason one machine cannot access the internet anymore, so you do whatever with it, but also close Chihaya. That's the gonna purge all the state in redis and renders the second tracker useless in the process.

I know we should clear our stuff out of redis at some point, but I'm not sure where. Comments @pushrax @jzelinskie ?

Reply:
Maybe we can use a redis counter to record the running chihaya, only the last exiting chihaya can clear the data.

Reply:
I don't think any chihaya instance should ever clear redis.

Reply:
@jzelinskie If we use Redis-A first and swtich to Redis-B, who is responsible for cleaning chihaya data in Redis-A.

Reply:
Hmm, we could document that we never clear redis, and document which keys are stored in redis, so the user can clean this up. Or we could add a small utility to do just that.

Reply:
Fine, simple but secure